src/main/java/org/candlepin/resource/RhsmApiCompatResource.java

```diff
@@ -0,0 +1,141 @@
+    @Transactional
+    public Stream<ConsumerEntitlementCountsDTO> listConsumerEntitlementCounts(
+        @Valid @NotNull RhsmApiConsumerEntitlementCountsQueryDTO rhsmApiConsumerEntitlementCountsQueryDTO,
+        Integer page, Integer perPage) {
+
+        if (page != null && page < 1) {
+            throw new BadRequestException(i18n.tr("Page query parameter must be greater than 0."));
+        }
+
+        if (perPage != null && (perPage < 1 || perPage > this.pageLimit)) {
+            throw new BadRequestException(i18n
+                .tr("Per page query parameter value must be larger than 0 and smaller than or equal to {0}.",
+                this.pageLimit));
+        }
+
+        List<String> consumerIds = rhsmApiConsumerEntitlementCountsQueryDTO.getConsumerIds();
+        List<String> consumerUuids = rhsmApiConsumerEntitlementCountsQueryDTO.getConsumerUuids();
+
+        List<ConsumerEntitlementCount> consumerEntCounts = this.rhsmApiCompatCurator
+            .getConsumerEntitlementCounts(consumerIds, consumerUuids);
+
+        // The order of the ConsumerEntitlementCountsDTOs needs to be in the same order as consumerEntCounts
+        List<ConsumerEntitlementCountsDTO> consumerEntCountDTOs = new ArrayList<>();
+        ConsumerEntitlementCountsDTO current = null;
+        for (ConsumerEntitlementCount consumerEntCount : consumerEntCounts) {
+            // Base case for the first iteration
+            if (current == null) {
+                current = new ConsumerEntitlementCountsDTO()
+                    .consumerId(consumerEntCount.id())
+                    .consumerUuid(consumerEntCount.uuid());
+            }
+
+            // We are now listing entitlement counts for a different consumer, so we can start populating
+            // a new ConsumerEntitlementCountsDTO for this consumer
+            if (!current.getConsumerId().equals(consumerEntCount.id())) {
+                consumerEntCountDTOs.add(current);
+
+                current = new ConsumerEntitlementCountsDTO()
+                    .consumerId(consumerEntCount.id())
+                    .consumerUuid(consumerEntCount.uuid());
+            }
+
+            RhsmApiEntitlementCountDTO entitlementCountsItem =
+                new RhsmApiEntitlementCountDTO()
+                    .contractNumber(consumerEntCount.contractNumber())
+                    .subscriptionId(consumerEntCount.subscriptionId())
+                    .productId(consumerEntCount.productId())
+                    .productName(consumerEntCount.productName())
+                    .count((int) consumerEntCount.quantity());
+
+            current.addEntitlementCountsItem(entitlementCountsItem);
+        }
+
+        // Add the last item
+        if (current != null) {
+            consumerEntCountDTOs.add(current);
+        }
+
+        ResteasyContext.pushContext(PageRequest.class, new PageRequest()
+            .setPage(page)
+            .setPerPage(perPage)
+            .setSortBy("consumerId")
```

Comment:
As mentioned above, ideally only the sort and order by adjustments are necessary, and those can be performed in the same place as the validation.

Reply:
Updated